### PR TITLE
Support multiple constructors

### PR DESCRIPTION
**Pull request by aalmada**
Looks for the constructor with all parameters corresponding to properties.
All tests are green.

### CODE REVIEW

File: Tests/WeaverTests.cs [hunk 1]
```diff
@@ -156,7 +170,18 @@ public void UniquePropertyTypeButExplicitMethodName_WithIsEmittedWithLongName()
...
+            m.IsPublic &&
+            m.Name == "With" &&
+            m.IsGenericMethod &&
+            m.GetParameters().Length == 1));
+        // check for a With method with Object parameter
+        Assert.False(type1.GetMethods().Any(m => 
+            m.IsPublic &&
+            m.Name == "With" && 
+            m.GetParameters().Length == 1 &&
+            m.GetParameters()[0].ParameterType == typeof(object)));
```
Comment (mikhailshilkov) [inline Tests/WeaverTests.cs]: I believe this change is not related to multiple constructor support. So,
a) Remove it from this PR and add a new PR if you need it.
b) Provide some motivation of why both generic and object need to be supported.
c) Add test cases.

File: With.Fody/ModuleWeaver.cs [hunk 2]
```diff
@@ -20,52 +21,57 @@ public ModuleWeaver()
 
     public void Execute()
     {
-        foreach (var type in ModuleDefinition.Types.Where(CanHaveWith))
+        foreach (var type in ModuleDefinition.Types)
         {
-            AddWith(type);
-            RemoveGenericWith(type);
-            LogInfo($"Added method 'With' to type '{type.Name}'.");
+            if (type.GetMethods().Any(m => m.IsPublic && m.Name.StartsWith("With")))
```
Comment (mikhailshilkov) [inline With.Fody/ModuleWeaver.cs]: Small thing, but I prefer `.Where` inside `foreach` to nested `if`. Saves you some indentation and reduces visual clutter.

File: With.Fody/ModuleWeaver.cs [hunk 3]
```diff
@@ -20,52 +21,57 @@ public ModuleWeaver()
...
+        return type.GetConstructors()
+            .Where(ctor =>
+                ctor.Parameters.Count >= 2 &&
+                ctor.Parameters.All(par =>
+                    type.Properties.Any(pro =>
+                        string.Compare(par.Name, pro.Name, StringComparison.InvariantCultureIgnoreCase) == 0
+                    )
+                )
+            )
+            .FirstOrDefault();
```
Comment (mikhailshilkov) [inline With.Fody/ModuleWeaver.cs]: This `.FirstOrDefault()` disturbs me... Now the outcome of the weaving might depend on constructor order, which is not good.
We find a first constructor with all parameters with names matching the properties. First one might have 2 parameters, but the second one might have 5. Should we prefer the later? That's why I didn't go for multiple constructors yet.
What is your use case?
More tests are needed for such scenarios.

File: With.Fody/ModuleWeaver.cs [hunk 4]
```diff
@@ -20,52 +21,57 @@ public ModuleWeaver()
...
     private void RemoveGenericWith(TypeDefinition type)
     {
-        foreach (var method in type.GetMethods().Where(m => m.IsPublic && m.Name == "With" && m.HasGenericParameters).ToArray())
+        var method = type.GetMethods()
+            .Where(m =>
+                m.IsPublic &&
+                m.Name == "With" &&
+                m.Parameters.Count == 1 &&
+                m.Parameters[0].ParameterType.FullName == "System.Object")
+            .FirstOrDefault();
```
Comment (mikhailshilkov) [inline With.Fody/ModuleWeaver.cs]: Again, not related to multi constructors.
Why do we remove just one?
Why `HasGenericParameters` replaced by `FullName == System.Object`?
Method name does not match the body anymore.

File: Tests/WeaverTests.cs (hunk 1 above)
Comment (aalmada) [inline Tests/WeaverTests.cs]: While testing the code for this PR, I noticed most test cases had an extra With. I thought RemoveGenericWith() was broken so I starting rewriting it. I then realised the issue was that most test cases have
```csharp
public MultipleConstructors With(object value) => this;
```
instead of the generics suggested in the documentation.
I first thought this was a valid alternative but now I realized that it may be an issue for classes with properties of type System.Object.
I'm removing this portion of the code but the test cases should be fixed.
instead of the

File: With.Fody/ModuleWeaver.cs (hunk 2 above)
Comment (aalmada) [inline With.Fody/ModuleWeaver.cs]: I also prefer _.Where_ but habits are hard to die... ;)
I'm changing it.

File: With.Fody/ModuleWeaver.cs (hunk 4 above)
Comment (aalmada) [inline With.Fody/ModuleWeaver.cs]: Answer is above.

File: Tests/WeaverTests.cs (hunk 1 above)
Comment (mikhailshilkov) [inline Tests/WeaverTests.cs]: Ah, that's my mistake then. The first version was using `object`, and then I switched to generic methods. I'll fix the tests, or you may submit a PR ;)

File: With.Fody/ModuleWeaver.cs (hunk 3 above)
Comment (aalmada) [inline With.Fody/ModuleWeaver.cs]: This query looks for constructors with the exact number of parameters as properties. For the case you mention, only the 5 parameter constructor would be returned.
The only case where more that one constructor would be found is if they had the same number of parameters but in a different order. I've never seen this case in the real world. It would be strange to have constructors behaving differently based on the order of parameters.
The query also checks for the names and I'm now adding type checking.
_With.Fody_ should allow any number of constructors as long as there is one that can be injected into the _.With_ methods. _GetValidConstructor()_ finds exactly that one, ignoring all the others.
I'm using _With.Fody_ on a large project and some more complex scenarios appear. Thanks again for developing this module as it's saving me a lot of trouble.

Comment (aalmada) [inline With.Fody/ModuleWeaver.cs]: Well, it least this was my intention. I'm writing more test and I'm finding issues... :(
I can't fix these right now but I hope to be back soon.

Comment (aalmada) [inline With.Fody/ModuleWeaver.cs]: Ended up fixing it. Please check.

**Comment by aalmada** [comment]
I'm done for now until further comments.
